packages/proxy/lib/http/util/prerequests.ts

```diff
@@ -12,101 +11,86 @@ const metrics: any = {
+          delete this.prerequestTimeouts[key]
+        }
+      })
+    }, 2000)
+  }
 
-      return cb(pendingBrowserPreRequest)
-    }
+  addPending (browserPreRequest: BrowserPreRequest) {
+    metrics.browserPreRequestsReceived++
+    const key = `${browserPreRequest.method}-${browserPreRequest.url}`
 
-    const timeout = setTimeout(() => {
-      metrics.neverReceivedPreRequest.push({ url: req.proxiedUrl })
-      ctxDebug('500ms passed without a pre-request, continuing request with an empty pre-request field!')
-
-      remove()
-      cb()
-    }, 500)
-
-    const startedMs = Date.now()
-    const remove = _.once(() => removeOne(this.requestsPendingPreRequestCbs, (v) => v === requestPendingPreRequestCb))
-
-    const requestPendingPreRequestCb = {
-      cb: (browserPreRequest) => {
-        const afterMs = Date.now() - startedMs
-
-        metrics.eventuallyReceivedPreRequest.push({ url: browserPreRequest.url, afterMs })
-        ctxDebug('received pre-request after %dms %o', afterMs, browserPreRequest)
-        clearTimeout(timeout)
-        remove()
-        cb(browserPreRequest)
-      },
-      proxiedUrl: req.proxiedUrl,
-      method: req.method,
+    if (this.pendingRequests[key]) {
+      debugVerbose('immediately matched pre-request %o', browserPreRequest)
+      clearTimeout(this.pendingRequests[key].timeout)
+      this.pendingRequests[key].callback(browserPreRequest)
+      delete this.pendingRequests[key]
     }
 
-    this.requestsPendingPreRequestCbs.push(requestPendingPreRequestCb)
+    debugVerbose('queuing pre-request to be matched later: %o', browserPreRequest)
+    this.pendingPreRequests[key] = browserPreRequest
+    this.prerequestTimeouts[key] = Date.now() + 10000
   }
 
-  addPending (browserPreRequest: BrowserPreRequest) {
-    if (this.pendingBrowserPreRequests.indexOf(browserPreRequest) !== -1) {
-      return
-    }
-
-    metrics.browserPreRequestsReceived++
+  get (req: CypressIncomingRequest, ctxDebug, callback: GetPreRequestCb) {
+    metrics.proxyRequestsReceived++
+    const key = `${req.method}-${req.proxiedUrl}`
 
-    const requestPendingPreRequestCb = removeOne(this.requestsPendingPreRequestCbs, (req) => {
-      return matches(browserPreRequest, req)
-    })
+    if (this.pendingPreRequests[key]) {
+      metrics.immediatelyMatchedRequests++
+      ctxDebug('matches pending pre-request %o', this.pendingPreRequests[key])
+      callback(this.pendingPreRequests[key])
 
-    if (requestPendingPreRequestCb) {
-      debugVerbose('immediately matched pre-request %o', browserPreRequest)
+      delete this.pendingPreRequests[key]
+      delete this.prerequestTimeouts[key]
 
-      return requestPendingPreRequestCb.cb(browserPreRequest)
+      return
     }
 
-    debugVerbose('queuing pre-request to be matched later %o %o', browserPreRequest, this.pendingBrowserPreRequests)
-    this.pendingBrowserPreRequests.push(browserPreRequest)
+    const timeout = setTimeout(() => {
```

Comment:
So there are two timeouts here: This is the timeout for unmatched *requests*, while the constructor timeout is to remove unmatched *prerequests*.

When a request times out (500ms, as in the existing code), we invoke its callback so that the server responds to the request.

When a prerequest times out, we don't need to do anything - we just clear up our reference to it.